Add benevolentUnionTypes (#8234)
phpstan/phpstan: website/src/config-reference.md

When set to `true`, PHPStan is strict about values with an unspecified (implicit `mixed`) type. It enables the same checks for values with no type specified that rule level 9 enables for explicitly specified `mixed` type values.

### `checkBenevolentUnionTypes`

<div class="text-xs inline-block border border-green-600 text-green-600 bg-green-100 rounded px-1 mb-4">Available in PHPStan 1.9.0</div>

**default**: `false`

> PHPStan has defined some benevolent union types. Those unions are not checked by the `checkUnionTypes` option enabled on level 7:

Reviewer: does the documentation mention anywhere what a "benevolent union type" actually is? if so it would be great to link to it here.

Reviewer: you should always prefer active voice when writing documentation:
https://www.grammarly.com/blog/active-vs-passive-voice/

```suggestion
PHPStan defines some benevolent union types, such as `array-key`. Benevolent unions are ignored by the `checkUnionTypes` option enabled on level 7:
```

Author: > does the documentation mention anywhere what a "benevolent union type" actually is? if so it would be great to link to it here.

There is none

Reviewer 2: this is one of the big open mysteries :D the best we have so far is this https://github.com/phpstan/phpstan/discussions/6440
I think it's a bit hard to understand and get and therefore mostly an internal thing to avoid annoyances

but, yeah, with this documentation it's not so internal anymore I guess :) so maybe somebody can describe them somehow

```php
public function requireInt(int $value): void {}
public function requireString(string $value): void {}

/**
 * @param array-key  $value1 // array-key is a benevolent union (int|string)
 * @param int|string $value2
 */
public function test(int|string $value1, int|string $value2): int
{
    $this->requireInt($value1);    // No error
    $this->requireString($value1); // No error
    $this->requireInt($value2);
    $this->requireString($value2);
}
```

> If you want a stricter analysis with union types, you can enable this option:

Reviewer: another case of active vs passive voice:

```suggestion
Enable stricter analysis of benevolent union types with the `checkBenevolentUnionTypes` option:
```
